### ISSUE TITLE
LDS ACK ERROR: sent weighted cluster with empty list 

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#14414]
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: infra-sl
  namespace: opsdemo
spec:
  hosts:
  - myhost.tcp.svc
  addresses:
  - 192.168.0.123/24
  ports:
  - name: tcp
    number: 8888
    protocol: TCP
  location: MESH_EXTERNAL
  resolution: STATIC
  endpoints:
  - address: 192.168.0.123
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: infra-sl
  namespace: opsdemo
spec:
  hosts:
  - myhost.tcp.svc
  gateways:
  - mesh
  - istio-egressgateway
  tcp:
  - match:
    - gateways:
      - mesh
      destinationSubnets:
      - 192.168.0.123/24
      port: 8888
    # route:
    # - destination:
    #     host: istio-egressgateway.istio-system.svc.cluster.local
    #     subset: infra-sl
    #     port:
    #       number: 9999
  - match:
    - gateways:
      - istio-egressgateway
      port: 9999
    route:
    - destination:
        host: myhost.tcp.svc
        port:
          number: 8888
      weight: 100
```

Gives

```
ADS:LDS: ACK ERROR 10.60.4.30:59316 sidecar~10.60.4.30~shell-6f85b78468-gjxfz.opsdemo~opsdemo.svc.cluster.local-22 (shell-6f85b78468-gjxfz.opsdemo) version_info:"2019-05-29T01:10:27Z/11" node:<id:"sidecar~10.60.4.30~shell-6f85b78468-gjxfz.opsdemo~opsdemo.svc.cluster.local" cluster:"shell.opsdemo" metadata:<fields:<key:"CONFIG_NAMESPACE" value:<string_value:"opsdemo" > > fields:<key:"INCLUDE_INBOUND_PORTS" value:<string_value:"" > > fields:<key:"INTERCEPTION_MODE" value:<string_value:"REDIRECT" > > fields:<key:"ISTIO_META_INSTANCE_IPS" value:<string_value:"10.60.4.30,10.60.4.30,fe80::50d3:32ff:fe9d:3783" > > fields:<key:"ISTIO_PROXY_SHA" value:<string_value:"istio-proxy:542b4fda35046fc9dad5014dbe81f3d5969e63f7" > > fields:<key:"ISTIO_PROXY_VERSION" value:<string_value:"1.1.3" > > fields:<key:"ISTIO_VERSION" value:<string_value:"master-20190522-09-15" > > fields:<key:"POD_NAME" value:<string_value:"shell-6f85b78468-gjxfz" > > fields:<key:"app" value:<string_value:"shell" > > fields:<key:"istio" value:<string_value:"sidecar" > > fields:<key:"pod-template-hash" value:<string_value:"6f85b78468" > > > locality:<region:"us-central1" zone:"us-central1-a" > build_version:"542b4fda35046fc9dad5014dbe81f3d5969e63f7/1.11.0-dev/Clean/RELEASE/BoringSSL" > type_url:"type.googleapis.com/envoy.api.v2.Listener" response_nonce:"5b9bd6f9-e0da-41f4-9093-660e540beccf" error_detail:<code:13 message:"Error adding/updating listener(s) 0.0.0.0_8888: Proto constraint validation failed (TcpProxyValidationError.WeightedClusters: [\"embedded message failed validation\"] | caused by WeightedClusterValidationError.Clusters: [\"value must contain at least \" '\\x01' \" item(s)\"]): stat_prefix: \"infra-sl.opsdemo\"\nweighted_clusters {\n}\n" >
```

One thing to note is this is failing the proto constraint. Maybe we can check the constraints pilot side.

Got this error from some modifications to https://github.com/istio/istio/issues/14334, so likely related

**Comment by howardjohn** [comment]
Root cause: https://github.com/istio/istio/blob/58c35be2433607a0451c1fe8ed9ec11270b52457/pilot/pkg/networking/core/v1alpha3/networkfilter.go#L122

never set routes since there are none.

Best fix is to reject this in Galley, need to follow up and see if it is invalid to provide no routes.

**Comment by howardjohn** [comment]
Very weird... we do actually validate? doesn't seem like it catches this for some reason though. https://github.com/istio/istio/blob/1d13d1f7e93c7d1433388da618573515ee0d1030/pilot/pkg/proxy/envoy/v2/lds.go#L59-L67

**Comment by howardjohn** [comment]
Action items:
* Need to determine we validation isn't happening, or if it is why it doesn't catch this.
* Once we fix validation we probably shouldn't panic...
* Need to determine if route is required or not
  * if it is, make it fail validation if missing
  * if it isn't, fix this bug + probably other bugs for http

**Comment by howardjohn** [comment]
Dug into this a bit

* Validation doesn't occur because this stuff is converted to a struct
* If we wanted to we can validate anytime we convert to struct. Not sure this is a good idea though

**Comment by howardjohn** [comment]
Waiting on https://github.com/istio/istio/pull/14726 landing in 1.2. Then we can bump this to P1 for the remaining action items (probably just open a new issue actually) for the `.Validate()` stuff
